osg-project-waittime/calculate-waittime.py

```diff
@@ -0,0 +1,235 @@
+            & Q("term", DN=user) 
+        ],
+    )
+    s = s.sort("StartTime")
+    #print(s.to_dict())
+    return s
+
+
+
+
+def getQueueTimes(users):
+    """
+    Loop through the raw usage to find the first 1000 hours during the interesting "days" for each user.
+    Calculate the average time in queue as (EndTime - QueueTime) - WallTime
+    """
+    # Create a mapping of user to userAttributes
+    userAttrDict = {}
+    for userAttr in users:
+        if userAttr.starttime < parser.parse("2021-03-09"):
+            #print("Before QueueTime available")
+            continue
+        found1000 = False
+        s = generateRawQuery(userAttr.username, userAttr.starttime, userAttr.endtime)
+        s = s.params(preserve_order=True)
+        for record in s.scan():
+            userAttr.njobs += 1
+            userAttr.walltime += record['WallDuration']
+            userAttr.corehours += record['CoreHours']
+            if 'QueueTime' in record:
+                userAttr.queuetime += (parser.parse(record['EndTime']).timestamp() - parser.parse(record['QueueTime']).timestamp()) - record['WallDuration']
+            else:
+                print("QueueTime not found when it should be for probe:{}")
+            # check if we have 1000 hours
+            if userAttr.corehours > 1000:
+                found1000 = True
+                break
+        if found1000:
+            print("Found the user: {} with QueueTime (hours): {} for CoreHours: {}".format(userAttr.username, userAttr.queuetime/HOUR, userAttr.corehours))
+        else:
+            print("Did not find 1000 hours in {} jobs of usage for user: {} in set of days: {}-{}".format(userAttr.njobs, userAttr.username, userAttr.starttime, userAttr.endtime))
+
+    return users
+
+
+
+def add_args():
+
+    argsparser = argparse.ArgumentParser(description='Calculate waittime for users')
+    argsparser.add_argument("starttime", type=str, help="Start Time, for example 2021-03-01")
+    argsparser.add_argument("endtime", type=str, help="End Time, for example 2021-03-31")
+    return argsparser
+
+def main():
+    argsparser = add_args()
+    args = argsparser.parse_args()
+
+    startDatetime = parser.parse(args.starttime)
+    endDatetime = parser.parse(args.endtime)
+
+    # Find all user with any usage in the last 1 month
+    perDay = getUsersPerDay(startDatetime, endDatetime)
+
+    # Gather per-day statistics for each user
+    users = getIdleUsers(perDay)
+
+    # Queue times
+    queueTimes = getQueueTimes(users)
+    columnNames = ["Username", "ProjectName", "Start Time", "End Time", "Days of Zero Usage", "Aggregate Hours In Queue", "Aggregate Core Hours", "Number of Jobs"]
+    df = pd.DataFrame(columns = columnNames)
+    i = 0
+    for userAttr in queueTimes:
+        df.loc[i] = [userAttr.username, ",".join(usernameToProject[userAttr.username]), userAttr.starttime, userAttr.endtime, userAttr.idledays, userAttr.queuetime/HOUR, userAttr.corehours, userAttr.njobs]
+        i+= 1
```

Comment:
Minor, but, are you doing anything tricky with `i` here, or can you use `enumerate` instead of explicitly initializing & incrementing it?

```suggestion
    for i, userAttr in enumerate(queueTimes):
        df.loc[i] = [userAttr.username, ",".join(usernameToProject[userAttr.username]), userAttr.starttime, userAttr.endtime, userAttr.idledays, userAttr.queuetime/HOUR, userAttr.corehours, userAttr.njobs]
```